### PR TITLE
feat(omen): plugin that integrates the decentralized prediction marke…

### LINKED ISSUE #3339: [Bounty] Integrate Omen, a prediction market into Mask plugins
## Catalogue

- Platform
  - [x] Gitcoin
  - [ ] Hackerlink
  - [ ] Mask
- Experience
  - [ ] Beginner
  - [x] Intermediate
  - [ ] Advanced
- Category
  - [ ] Mask Core Feature
  - [x] Mask Plugin
  - [ ] Security/Cryptography Feature
  - [ ] Documentation
  - [ ] Design
  - [ ] Other
    - [ ] /_List it here_/
- Time Commitment
  - [x] 2 weeks
- Bounty Size
  - [x] 400 $MASK or $2000

## Part 1: Problem Description

Prediction market is fun to play with friends so naturally it should live on social networks. Mask would love to see all sorts of prediction markets thrive on us. Notable ones including Augur, Omen (powered by gnosis), Polymarket, Reality card etc

The UI could resemble dhedge in a card style with different tabs serving different purposes

image
_Dhedge UI_

image
_Detail page for a market on Omen_

This detail page could serve as the default tab on mask omen plugin. With pool and history tab as the second and third tab. The current page looks a bit messy and less user friendly so it might need to be trimmed down a bit.

Omen has choosen xdai as the scaling solution so we will probably leave it be and focus on eth mainnet first.

## Part 2: Resources

- [Bounty Guide](https://github.com/DimensionDev/Maskbook/tree/master/docs/bounty-guide.md)
- https://omen.eth.link/#/liquidity
- https://github.com/protofire/omen-exchange

## Part 3: Contact and Support
@yisiliu for ideas
@Jack-Works for implementations
Mask Network discord server: https://discord.gg/4Xccj7a75Q
Omen discord server: discord.gg/4QXEJQkvHH

Submission Requirements
Please include demo to show the integration is working.

Winner Announcement Date
Winners will be announced at the conclusion of Gitcoin 10 Hackathon

Judging Criteria
Have a working demo.
Clean code & tech design
Clear UI, easy to use and generally good user experience

### PR DESCRIPTION
**Pull request by CriptKeeper**
## Description

Closes #3339
Plugin that integrates the decentralized prediction market platform Omen.

## Type of change
- [x] New feature (non-breaking change which adds functionality)

### Swap View
omen-view-swap

### Pool View
omen-view-pool

### History View
omen-view-history

## Checklist

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
  - [x] I have removed all in development `console.log`s
  - [x] I have removed all commented code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

### CODE REVIEW

File: packages/mask/src/plugins/Omen/UI/OmenDetailsCard.tsx [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    let sharesAmount = new BigNumber(0)
+    const liquidityArr = Object.values(liquidityData)
+    for (const liquidityObj of liquidityArr) {
+        const tempSum = new BigNumber(sharesAmount)
+        const nextVal = new BigNumber(liquidityObj.sharesAmount)
+        sharesAmount = tempSum.plus(nextVal)
+    }
+    const liquidityAmount = new BigNumber(sharesAmount).shiftedBy(collateralDecimals).toNumber().toPrecision(4)
+    const totalVolume = new BigNumber(collateralVolume).shiftedBy(collateralDecimals).toNumber().toPrecision(4)
+    const dailyVolume = Number(usdRunningDailyVolume).toPrecision(3)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Omen/UI/OmenDetailsCard.tsx]: Recommend to use `parseInt` or `parseFloat`.

File: packages/mask/src/plugins/Omen/UI/OmenDetailsCard.tsx [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    const tokenDataRes = useFetchToken(tokenId)
+    const tokenObj: tokenData | null | undefined = tokenDataRes ? tokenDataRes.value : undefined
+    const collateralToken = tokenObj?.registeredToken
+    const collateralDecimals = collateralToken ? collateralToken.decimals : 0
+    let sharesAmount = new BigNumber(0)
+    const liquidityArr = Object.values(liquidityData)
+    for (const liquidityObj of liquidityArr) {
+        const tempSum = new BigNumber(sharesAmount)
+        const nextVal = new BigNumber(liquidityObj.sharesAmount)
+        sharesAmount = tempSum.plus(nextVal)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Omen/UI/OmenDetailsCard.tsx]: Looks strange to me, only the last item of `liquidityArr` will take into account.

File: packages/mask/src/plugins/Omen/UI/OmenDetailsCard.tsx [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+export function OmenDetailsCard(props: OmenDetailsCardProps) {
+    const { t } = useI18N()
+    const { classes } = useStyles()
+    const { liquidityData, tokenId, usdRunningDailyVolume, closingTimestamp, collateralVolume } = props
+    const tokenDataRes = useFetchToken(tokenId)
+    const tokenObj: tokenData | null | undefined = tokenDataRes ? tokenDataRes.value : undefined
+    const collateralToken = tokenObj?.registeredToken
+    const collateralDecimals = collateralToken ? collateralToken.decimals : 0
+    let sharesAmount = new BigNumber(0)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Omen/UI/OmenDetailsCard.tsx]:
```suggestion
    let sharesAmount = ZERO
```

File: packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+    TableBody,
+    TableRow,
+    TableCell,
+    Paper,
+} from '@mui/material'
+import { useI18N } from '../../../utils'
+import { bigNumberToString } from '../utils'
+import { useFetchToken } from '../hooks/useToken'
+// import { PluginOmenMessages } from '../messages'
+// import { useRemoteControlledDialog } from '@masknet/shared'
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx]:
```suggestion
```

File: packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+    const tokenSymbol = collateralToken ? collateralToken.symbol : ''
+    const earnTradingFee = bigNumberToString(marketFee.multipliedBy(Math.pow(10, 2)), STANDARD_DECIMALS)
+
+    // const totalPoolShares = marketLiquidity.sharesAmount
+    // const poolTokens = calcPoolTokens(
+    //   inputAmount || new BigNumber(0),
+    //   balances.map(b => b.holdings),
+    //   totalPoolShares,
+    // )
+
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx]:
```suggestion
```

File: packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+    marketOutcomes: string[]
+    marketOutcomePrices: BigNumber[]
+    marketFee: BigNumber
+    tokenId: string
+}
+
+export function OmenPoolView(props: OmenPoolViewProps) {
+    const { t } = useI18N()
+    const { classes } = useStyles()
+    const [inputAmount, setInputAmount] = useState<BigNumber>(new BigNumber(0))
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx:126]:
```suggestion
    const [inputAmount, setInputAmount] = useState(ZERO)
```

File: packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+    const { marketId, marketOutcomes, marketOutcomePrices, tokenId } = props
+    const tokenDataRes = useFetchToken(tokenId)
+    const tokenObj: tokenData | null | undefined = tokenDataRes ? tokenDataRes.value : undefined
+    const marketFee = props.marketFee ? new BigNumber(props.marketFee) : new BigNumber(0)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx:130]:
```suggestion
    const marketFee = new BigNumber(props.marketFee ?? ZERO)
```

File: packages/mask/src/plugins/Omen/utils.ts [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        unitIndex += 1
+        rNumber = rNumber / 1000
+    }
+
+    return `${Number.parseFloat(rNumber.toFixed(decimals))}${units[unitIndex]}`
+}
+
+export function bigNumberToNumber(num: BigNumber | string, decimal = 8, fix = 8) {}
+
+export function bigNumberToString(amount: BigNumber, decimals: number) {
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Omen/utils.ts:54]: Does `formatBalance` work for you?

File: packages/mask/src/plugins/Omen/utils.ts [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+    while (rNumber >= 1000 && unitIndex < 5) {
+        unitIndex += 1
+        rNumber = rNumber / 1000
+    }
+
+    return `${Number.parseFloat(rNumber.toFixed(decimals))}${units[unitIndex]}`
+}
+
+export function bigNumberToNumber(num: BigNumber | string, decimal = 8, fix = 8) {}
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Omen/utils.ts]:
```suggestion
```

File: packages/mask/src/plugins/Omen/utils.ts [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    FORMAT: normalFormat,
+    ROUNDING_MODE: BigNumber.ROUND_FLOOR,
+}
+
+const groupConfig = {
+    FORMAT: groupFormat,
+    ROUNDING_MODE: BigNumber.ROUND_FLOOR,
+}
+
+BigNumber.config(normalConfig)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Omen/utils.ts:32]: This behavior will change global settings. Please avoid doing this.

File: packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+            // openPoolDialog({
+            //     open: true,
+            //     marketId: marketId,
+            //     txType: txType,
+            //     inputAmount: inputAmount,
+            // })
+        },
+        [marketId, inputAmount],
+    )
+    // #endregion
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx:161]: For debugging?

File: packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+                            <Grid container item direction="column" alignItems="center" className={classes.buttonCol}>
+                                <Grid item alignSelf="left" xs={4} textAlign="center">
+                                    <Button
+                                        className={classes.button}
+                                        variant="outlined"
+                                        fullWidth
+                                        color="primary"
+                                        href={`${OMEN_BASE_URL}${marketId}/pool`}
+                                        target="_blank"
+                                        /* onClick={() => onPoolRequest('deposit')} */
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx]:
```suggestion
```

File: packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+                                </Grid>
+                                <Grid item alignSelf="right" xs={4} textAlign="center">
+                                    <Button
+                                        className={classes.button}
+                                        variant="outlined"
+                                        fullWidth
+                                        color="primary"
+                                        href={`${OMEN_BASE_URL}${marketId}/pool`}
+                                        target="_blank"
+                                        /* onClick={() => onPoolRequest('withdraw')} */
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx]:
```suggestion
```

File: packages/mask/src/plugins/Omen/UI/OmenSwapView.tsx [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+    Table,
+    TableBody,
+    TableRow,
+    TableCell,
+    TableHead,
+    LinearProgress,
+} from '@mui/material'
+import { useI18N } from '../../../utils'
+// import { PluginOmenMessages } from '../messages'
+// import { useRemoteControlledDialog } from '@masknet/shared'
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Omen/UI/OmenSwapView.tsx]:
```suggestion
```

File: packages/mask/src/plugins/Omen/UI/OmenSwapView.tsx [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+            console.log('onSwapRequest marketData: ', marketId)
+            // openSwapDialog({
+            //     open: true,
+            //     marketId: marketId,
+            //     txType: txType,
+            // })
+        },
+        [marketId],
+    )
+    // #endregion
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Omen/UI/OmenSwapView.tsx:110]: For debugging?

File: packages/mask/src/plugins/Omen/UI/OmenSwapView.tsx [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+            <Grid container item direction="row" className={classes.buttonRow}>
+                <Grid item xs={4} textAlign="center">
+                    <Button
+                        className={classes.button}
+                        variant="contained"
+                        fullWidth
+                        color="primary"
+                        href={`${OMEN_BASE_URL}${marketId}/buy`}
+                        target="_blank"
+                        /* onClick={() => onSwapRequest('buy')} */
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Omen/UI/OmenSwapView.tsx]:
```suggestion
```

File: packages/mask/src/plugins/Omen/UI/OmenSwapView.tsx [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+                </Grid>
+                <Grid item xs={4} textAlign="center">
+                    <Button
+                        className={classes.button}
+                        variant="contained"
+                        fullWidth
+                        color="primary"
+                        href={`${OMEN_BASE_URL}${marketId}/sell`}
+                        target="_blank"
+                        /* onClick={() => onSwapRequest('sell')} */
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Omen/UI/OmenSwapView.tsx]:
```suggestion
```

File: packages/mask/src/plugins/Omen/UI/OmenView.tsx [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+
+    return (
+        <Box className={classes.root}>
+            <Typography className={classes.title}>{marketData.title}</Typography>
+
+            <OmenDetailsCard
+                liquidityData={liquidityData}
+                tokenId={marketData.collateralToken}
+                usdRunningDailyVolume={marketData.usdRunningDailyVolume}
+                closingTimestamp={new Date(1000 * +marketData.creationTimestamp)}
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Omen/UI/OmenView.tsx:104]: Recommend to use `fromUnixTime`.

File: packages/mask/src/plugins/Omen/hooks/useHistory.ts [hunk 19]
```diff
@@ -0,0 +1,10 @@
+import { useAsyncRetry } from 'react-use'
+import { fetchHistory } from '../apis'
+
+export function useFetchHistory(id: string) {
+    return useAsyncRetry(async () => {
+        if (!id) return null
+
+        return fetchHistory(id)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Omen/hooks/useHistory.ts:8]: Recommend to use service RPC call.

File: packages/mask/src/plugins/Omen/hooks/useMarket.ts [hunk 20]
```diff
@@ -0,0 +1,10 @@
+import { useAsyncRetry } from 'react-use'
+import { fetchMarket } from '../apis'
+
+export function useFetchMarket(id: string) {
+    return useAsyncRetry(async () => {
+        if (!id) return null
+
+        return fetchMarket(id)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Omen/hooks/useMarket.ts:8]: Recommend to use services RPC call.

File: packages/mask/src/plugins/Omen/hooks/useToken.ts [hunk 21]
```diff
@@ -0,0 +1,10 @@
+import { useAsyncRetry } from 'react-use'
+import { fetchToken } from '../apis'
+
+export function useFetchToken(id: string) {
+    return useAsyncRetry(async () => {
+        if (!id) return null
+
+        return fetchToken(id)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Omen/hooks/useToken.ts:8]: Recommend to use services RPC call.

File: packages/mask/src/plugins/Omen/README.md [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+
+## Feature Set
+
+- [x] Swap
+- [x] Pool
+- [x] History
+
+## Related Projects
+
+- <https://omen.eth.link/#/>
```
Comment (septs) [inline packages/mask/src/plugins/Omen/README.md]:
```suggestion
- <https://omen.eth.link>
```

File: packages/mask/src/plugins/Omen/README.md [hunk 23]
```diff
@@ -0,0 +1,22 @@
+# Plugin: Omen
+
+Plugin that integrates the decentralized prediction market platform Omen.
+Developed by @Dxdao\_ and powered by the Gnosis conditional tokens framework.
```
Comment (septs) [inline packages/mask/src/plugins/Omen/README.md]:
```suggestion
Developed by `@Dxdao_` and powered by the Gnosis conditional tokens framework.
```

File: packages/mask/src/plugins/Omen/README.md [hunk 24]
```diff
@@ -0,0 +1,22 @@
...
+
+- <https://omen.eth.link/#/>
+
+## Related Discussion
+
+- <https://github.com/DimensionDev/Maskbook/issues/3339>
+
+## Pull Requests
+
+- <>
```
Comment (septs) [inline packages/mask/src/plugins/Omen/README.md]:
```suggestion
- <https://github.com/DimensionDev/Maskbook/pull/5844>
```

File: packages/mask/src/plugins/Omen/SNSAdaptor/index.tsx [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+import { parseURL } from '@masknet/shared-base'
+import { ChainId } from '@masknet/web3-shared-evm'
+import { extractTextFromTypedMessage } from '@masknet/typed-message'
+import { Plugin, usePluginWrapper, usePostInfoDetails } from '@masknet/plugin-infra'
+import { base } from '../base'
+import { OmenView } from '../UI/OmenView'
+import { EthereumChainBoundary } from '../../../web3/UI/EthereumChainBoundary'
+
+const regexLinkPattern = /^https:\/\/omen.eth.link\/#\/(0x\w+)/
+const isOmenURL = (link: string): boolean => regexLinkPattern.test(link)
```
Comment (septs) [inline packages/mask/src/plugins/Omen/SNSAdaptor/index.tsx]:
```suggestion
const isOmenURL = (link: string) => regexLinkPattern.test(link)
```

File: packages/mask/src/plugins/Omen/UI/OmenDetailsCard.tsx (hunk 2 above)
Comment (septs) [inline packages/mask/src/plugins/Omen/UI/OmenDetailsCard.tsx]:
```suggestion
    const sharesAmount = BigNumber.sum(...Object.values(liquidityData).map((item) => item.sharesAmount))
```

File: packages/mask/src/plugins/Omen/UI/OmenDetailsCard.tsx [hunk 26]
```diff
@@ -0,0 +1,121 @@
...
+    const liquidityArr = Object.values(liquidityData)
+    for (const liquidityObj of liquidityArr) {
+        const tempSum = new BigNumber(sharesAmount)
+        const nextVal = new BigNumber(liquidityObj.sharesAmount)
+        sharesAmount = tempSum.plus(nextVal)
+    }
+    const liquidityAmount = new BigNumber(sharesAmount).shiftedBy(collateralDecimals).toNumber().toPrecision(4)
+    const totalVolume = new BigNumber(collateralVolume).shiftedBy(collateralDecimals).toNumber().toPrecision(4)
```
Comment (septs) [inline packages/mask/src/plugins/Omen/UI/OmenDetailsCard.tsx]:
```suggestion
    const liquidityAmount = sharesAmount.shiftedBy(collateralDecimals).toPrecision(4)
    const totalVolume = collateralVolume.shiftedBy(collateralDecimals).toPrecision(4)
```

File: packages/mask/src/plugins/Omen/UI/OmenHistoryView.tsx [hunk 27]
```diff
@@ -0,0 +1,365 @@
...
+                                        </TableCell>
+                                        <TableCell>
+                                            <Typography className={classes.tCellLeft}>{row.transactionType}</Typography>
+                                        </TableCell>
+                                        <TableCell>
+                                            <Typography className={classes.tCellCenter}>
+                                                {new BigNumber(row.sharesOrPoolTokenAmount)
+                                                    .shiftedBy(collateralDecimals)
+                                                    .toNumber()
+                                                    .toPrecision(2)}
```
Comment (septs) [inline packages/mask/src/plugins/Omen/UI/OmenHistoryView.tsx]:
```suggestion
                                                {row.sharesOrPoolTokenAmount.shiftedBy(collateralDecimals).toPrecision(2)}
```

File: packages/mask/src/plugins/Omen/UI/OmenHistoryView.tsx [hunk 28]
```diff
@@ -0,0 +1,365 @@
...
+                                            </Typography>
+                                        </TableCell>
+                                        <TableCell>
+                                            <Typography className={classes.tCellRight}>
+                                                {new BigNumber(row.collateralTokenAmount)
+                                                    .shiftedBy(collateralDecimals)
+                                                    .toNumber()
+                                                    .toPrecision(2)}{' '}
```
Comment (septs) [inline packages/mask/src/plugins/Omen/UI/OmenHistoryView.tsx]:
```suggestion
                                                {row.collateralTokenAmount.shiftedBy(collateralDecimals).toPrecision(2)}{' '}
```

File: packages/mask/src/plugins/Omen/UI/OmenHistoryView.tsx [hunk 29]
```diff
@@ -0,0 +1,365 @@
...
+                                                {tokenSymbol}
+                                            </Typography>
+                                        </TableCell>
+                                        <TableCell>
+                                            <Typography className={classes.tCellRight}>
+                                                {new Date(keys[index]).toDateString()}
```
Comment (septs) [inline packages/mask/src/plugins/Omen/UI/OmenHistoryView.tsx:307]: @guanbinrui @Jack-Works suggest, update documentation specify date-time format method

e.g:
```js
import formatDateTime from "date-fns/format"

{formatDateTime(input, 'yyyy-MM-dd')}
```

File: packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx [hunk 30]
```diff
@@ -0,0 +1,323 @@
...
+        paddingLeft: theme.spacing(4),
+        '&:hover': {
+            backgroundColor: 'white',
+            color: '#212529',
+            border: '1px solid #393b4a',
+        },
+    },
+    headerLeft: {
+        color: MaskColorVar.normalText,
+        fontWeight: theme.typography.fontWeightRegular as any,
```
Comment (septs) [inline packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx:54]: plz do not use `any` type.

File: packages/mask/src/plugins/Omen/apis/index.ts [hunk 31]
```diff
@@ -0,0 +1,191 @@
...
+
+let cachedMarketLiquidity: fpmmLiquidityData | undefined
+let marketLiquidityLastCachedAt = Date.now()
+let marketLiquidityLastIdCached: string
+export async function fetchLiquidity(id: string) {
+    // Check if it's cached first
+    if (
+        marketLiquidityLastIdCached === id &&
+        cachedMarketLiquidity &&
+        Date.now() - marketLiquidityLastCachedAt > 30 * 1000
```
Comment (septs) [inline packages/mask/src/plugins/Omen/apis/index.ts:119]: plz use https://date-fns.org/v2.28.0/docs/isBefore

File: packages/mask/src/plugins/Omen/apis/index.ts [hunk 32]
```diff
@@ -0,0 +1,191 @@
...
+        }`,
+    }
+
+    const response = await fetch(OMEN_SUBGRAPH_URL, {
+        body: JSON.stringify(body),
+        method: 'POST',
+        headers: { 'Content-Type': 'application/json' },
+    })
+
+    cachedMarketDetails = (await response.json())?.data
```
Comment (septs) [inline packages/mask/src/plugins/Omen/apis/index.ts:64]: plz use https://github.com/DimensionDev/kit/blob/master/src/ecmascript/memoizePromise.ts

```ts
import { memoizePromise } from "@dimensiondev/kit"
```

**Review by septs — CHANGES_REQUESTED** [review]

**Comment by UncleBill** [comment]
Hi @CriptKeeper , the branch behind the main branch a lot, please sync with it. Do not mix in so much commits from the main branch.

File: packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx [hunk 33]
```diff
@@ -0,0 +1,312 @@
...
+                    <Grid container item direction="row" wrap="nowrap" justifyItems="center">
+                        <Grid item>
+                            <Grid container item direction="column" alignItems="center" className={classes.buttonCol}>
+                                <Grid item alignSelf="left" xs={4} textAlign="center">
+                                    <Button
+                                        className={classes.button}
+                                        variant="outlined"
+                                        fullWidth
+                                        color="primary"
+                                        href={`${OMEN_BASE_URL}${marketId}/pool`}
```
Comment (septs) [inline packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx:214]: plz use https://npm.im/urlcat build the url

File: packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx [hunk 34]
```diff
@@ -0,0 +1,312 @@
...
+                                        {t('plugin_omen_deposit')}
+                                    </Button>
+                                </Grid>
+                                <Grid item alignSelf="right" xs={4} textAlign="center">
+                                    <Button
+                                        className={classes.button}
+                                        variant="outlined"
+                                        fullWidth
+                                        color="primary"
+                                        href={`${OMEN_BASE_URL}${marketId}/pool`}
```
Comment (septs) [inline packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx:225]: plz use https://npm.im/urlcat build the url

File: packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx [hunk 35]
```diff
@@ -0,0 +1,312 @@
...
+                            </Grid>
+                        </Paper>
+                    </Grid>
+                </Grid>
+            </Grid>
+        </div>
+    )
+}
+
+// // Total Liquidity
```
Comment (septs) [inline packages/mask/src/plugins/Omen/UI/OmenPoolView.tsx:256]: plz remove all comments.

File: packages/mask/src/plugins/Omen/UI/OmenSwapView.tsx [hunk 36]
```diff
@@ -0,0 +1,199 @@
...
+                </TableBody>
+            </Table>
+            <Grid container item direction="row" className={classes.buttonRow}>
+                <Grid item xs={4} textAlign="center">
+                    <Button
+                        className={classes.button}
+                        variant="contained"
+                        fullWidth
+                        color="primary"
+                        href={`${OMEN_BASE_URL}${marketId}/buy`}
```
Comment (septs) [inline packages/mask/src/plugins/Omen/UI/OmenSwapView.tsx:178]: plz use https://npm.im/urlcat build the url

File: packages/mask/src/plugins/Omen/UI/OmenSwapView.tsx [hunk 37]
```diff
@@ -0,0 +1,199 @@
...
+                        {t('plugin_omen_buy')}
+                    </Button>
+                </Grid>
+                <Grid item xs={4} textAlign="center">
+                    <Button
+                        className={classes.button}
+                        variant="contained"
+                        fullWidth
+                        color="primary"
+                        href={`${OMEN_BASE_URL}${marketId}/sell`}
```
Comment (septs) [inline packages/mask/src/plugins/Omen/UI/OmenSwapView.tsx:189]: plz use https://npm.im/urlcat build the url

**Pull request closed without merging**